cime_config/SystemTests/mksurfdataesmf.py

```diff
@@ -0,0 +1,125 @@
+        self._TestStatus_log_path = os.path.join(self._get_caseroot(),
+            'TestStatus.log')
+
+    def build_phase(self, sharedlib_only=False, model_only=False):
+        """
+        Build executable that will generate fsurdat
+        Generate namelist for generating fsurdat
+        SKIP Generate jobscript that runs executable
+        Modify user_nl_clm to point to the generated fsurdat
+        """
+        # build_phase gets called twice:
+        # - once with sharedlib_only = True and
+        # - once with model_only = True
+        # Call the following steps only once during the test but do not skip
+        # if the test stops and gets restarted.
+        if sharedlib_only:
+            # Paths and strings
+            build_script_path = os.path.join(self._tool_path,
+                'gen_mksurfdata_build.sh')
+            nml_script_path = os.path.join(self._tool_path,
+                'gen_mksurfdata_namelist.py')
+            gen_mksurfdata_namelist = f'{nml_script_path} --res {self._res} --start-year {self._model_yr} --end-year {self._model_yr}'
+
+            # Rm tool_bld and build executable that will generate fsurdat
+            try:
+                subprocess.check_call(f'rm -rf {self._tool_bld}', shell=True)
+            except subprocess.CalledProcessError as e:
+                sys.exit(f'{e} ERROR REMOVING {self._tool_bld}. DETAILS IN {self._TestStatus_log_path}')
+            try:
+                subprocess.check_call(f'{build_script_path} --blddir {self._tool_bld}', shell=True)
+            except subprocess.CalledProcessError as e:
+                sys.exit(f'{e} ERROR RUNNING {build_script_path}. DETAILS IN {self._TestStatus_log_path}')
+
+            # Generate namelist for generating fsurdat (rm namelist if exists)
+            if os.path.exists(self._fsurdat_namelist):
+                os.remove(self._fsurdat_namelist)
+            try:
+                subprocess.check_call(gen_mksurfdata_namelist, shell=True)
+            except subprocess.CalledProcessError as e:
+                sys.exit(f'{e} ERROR RUNNING {gen_mksurfdata_namelist}. DETAILS IN {self._TestStatus_log_path}')
+
+        # Call this step only once even if the test stops and gets restarted.
+        if not os.path.exists(os.path.join(self._get_caseroot(),
+            'done_MKSURFDATAESMF_setup.txt')):
+            # Modify user_nl_clm to point to the generated fsurdat
+            self._modify_user_nl()
+            with open('done_MKSURFDATAESMF_setup.txt', 'w') as fp:
+                pass
+
+        self.build_indv(sharedlib_only=sharedlib_only, model_only=model_only)
+
+    def run_phase(self):
+        """
+        Run executable to generate fsurdat
+        Submit CTSM run that uses fsurdat just generated
+        """
+        # Paths and command strings
+        executable_path = os.path.join(self._tool_bld, 'mksurfdata')
+        machine = self._case.get_value("MACH")
+        if machine == 'cheyenne':
```

Comment:
You should be able to refactor this a bit to rely on the gen_mksurfdata_jobscript_single.py script to produce the commands that need to be run rather than hardcoding the machines that can be used here. That would be a nice improvement, but this could also be a future change. The advantage is that it removes embedding of machine specific things into this script, and puts all the machine specific run into one place.

Reply:
...so run `gen_mksurfdata_jobscript_single.py` and then run the jobscript without qsub in front.
...and most general approach is to pick up the mpi command from `env_mach_specific.xml`